sklearn/preprocessing/knn_imputation.py

```diff
@@ -0,0 +1,253 @@
+            The input data to complete.
+        """
+        force_all_finite = False if self.missing_values in ["NaN",
+                                                            np.nan] else True
+        check_is_fitted(self, 'statistics_')
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        mask = _get_mask(X, self.missing_values)
+        n_rows_X, n_cols_X = X.shape
+        row_total_missing = mask.sum(axis=1)
+
+        # Get fitted objects
+        fitted_NNObj, fitted_col_means = self.statistics_
+        fitted_X = fitted_NNObj._fit_X
+
+        # Check for excessive missingness in rows
+        bad_rows = row_total_missing > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "The following rows have more than {0}% missing values and "
+                "are imputed with column means: {1}"
+                    .format(self.row_max_missing*100, np.where(bad_rows)))
+            X_bad = X[bad_rows, :]
+            X = X[~bad_rows, :]
+            mask = _get_mask(X, self.missing_values)
+            row_total_missing = mask.sum(axis=1)
+
+        # Check if the X in fit() and transform() are the same
+        if X is fitted_X or X.base is fitted_X.base:
+            neighbors = fitted_NNObj.kneighbors(n_neighbors=self.n_neighbors)
+        else:
+            neighbors = fitted_NNObj.kneighbors(X,
+                                                n_neighbors=self.n_neighbors)
+
+        # Get row index, distance, and weights of donors
+        knn_distances, knn_row_index = neighbors
+        weights = _get_weights(knn_distances[row_total_missing.astype(
+            np.bool), ], self.weights)
+
+        knn_row_index = np.vsplit(knn_row_index, knn_row_index.shape[0])
+        knn_row_index = np.repeat(knn_row_index,
+                                  row_total_missing, axis=0).ravel()
+
+        # Get column index of donors
+        # NOTE: Following assumes columns in X and _fit_X are in the same order
+        col_missing_index = np.where(mask)[1]
+        knn_col_index = np.repeat(col_missing_index, self.n_neighbors)
+
+        # Calculate kNN score and impute
+        donors = fitted_X[
+            (knn_row_index, knn_col_index)].reshape((-1, self.n_neighbors))
+        donors = np.ma.array(
+            donors, mask=_get_mask(donors, self.missing_values))
+        imputed = np.ma.average(donors, axis=1, weights=weights)
+        X[mask] = imputed.data
+
+        # Merge bad rows to X and mean impute any leftover missing
+        if np.any(bad_rows):
+            X_merged = np.empty((n_rows_X, n_cols_X))
+            X_merged[bad_rows, :] = X_bad
+            X_merged[~bad_rows, :] = X
+            X = X_merged
+
+        # Impute bad_rows and leftover missing with column means
+        mask_after_knn = _get_mask(X, self.missing_values)
```

Comment:
Ideally we'd only get the mask for rows where we know that values may still need imputation.